[quantization] Make observed values actually flow through observers

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30140 [quantization] Make observed values actually flow through observers**
* #30130 [quantization] Modernize graph mode IR API calls

This seems more semantically correct to me, and makes it so we don't have to iterate over Uses of observed values

Differential Revision: [D18610676](https://our.internmc.facebook.com/intern/diff/D18610676)

Commit:
[quantization] Make observed values actually flow through observers

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -226,7 +226,13 @@ Node* InsertObserversHelper::insertObserverFor(
+    call->output()->copyMetadata(v);
+
+    // Replace v with the output of observer
+    v->replaceAllUsesWith(call->output());
```
Do we require all observers to return unchanged input value? Is it how they are implemented now?

jerryzh168 replied:
Yes that is a contract we have

ZolotukhinM reviewed (approved):
Looks good!

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -452,7 +456,7 @@ class QuantizeHelper {
       Value* child_instance);
   void quantizeTensor(Value* v);
-  void removeObserver(Value* v, const std::string& observer_name);
+  Value* removeObserver(Value* v, const std::string& observer_name);
```
can you add some comments to this function? especially input Value and return value?

jerryzh168 reviewed (approved):
LG

Commit:
Update on "[quantization] Make observed values actually flow through observers"


This seems more semantically correct to me, and makes it so we don't have to iterate over Uses of observed values

Differential Revision: [D18610676](https://our.internmc.facebook.com/intern/diff/D18610676)

[ghstack-poisoned]
